### ISSUE TITLE
OddpubMetrics Docker image and Database Table Generation

### ISSUE THREAD

**Issue by joshlawrimore** [nimh-dsst/dsst-etl#13]
1. Create a docker image to run a service for the [oddpub tool](https://github.com/quest-bih/oddpub/tree/v7.1.0). [This example](https://github.com/nimh-dsst/osm/tree/main/external_components/rtransparent) runs rtransparent (which uses oddpub under the hood). Remove superfluous dependencies and adapt the API to run oddpub 7.1.0 instead
2. Create a schema for a table OddpubMetrics. This should match the fields that are written out by running the oddpub tool (v7.1.0) i.e. inspect the outputs of oddpub
3. Write a script taking into account the following: 
    a. It should take a path (or many) to input pdf(s). For each pdf run oddpub  and save the output and the intermediate  to a file. 
    b. Create and reference entries in the Work, Document, and Provenance, RTransparentMetrics  tables etc. as required.
    c. An intermediate between the pdf and the metrics should be captured as a Document
4. The work should be accompanied by an Alembic migration script
5. The work should include tests that write to a locally deployed postgres database (rollback each transaction)

**Comment by quang-ng** [comment]
Hello @joshlawrimore  and @leej3 , there are some questions regarding this issue:

1. At [the example](https://github.com/nimh-dsst/osm/tree/main/external_components/rtransparent), we have FastAPI. Should we develop API endpoints like this, or simply create a script to run as we did before?

2. In point 3.a, it mentions that `it should take a path (or multiple paths) to input PDF(s). For each PDF, run oddpub and save the output and the intermediate to a file.` Why is it necessary to save the output to a file? Shouldn't we just store it in the OddpubMetrics database table? If we do need to save to a file, will this happen every time?

3. Regarding point 3.c, which states that `an intermediate between the PDF and the metrics should be captured as a Document,` could you clarify what is meant by `intermediate between PDF and metrics`? What is it, and how should it be generated?

**Comment by leej3** [comment]
> Should we develop API endpoints like this, or simply create a script to run as we did before?

Yes the idea of encapsulating the pipeline as a service is to help isolate the dependencies across different pipelines. This one is not especially hard to run locally but we don't want to accumulate lots of conda environments. 

> Why is it necessary to save the output to a file? Shouldn't we just store it in the OddpubMetrics database table? If we do need to save to a file, will this happen every time?

Good point. Let's drop this requirement for now. Just upload to the table.

> could you clarify what is meant by intermediate between PDF and metrics? What is it, and how should it be generated?

In general the intermediate will be a text representation of the pdf. This is the output of the "parser" i.e. the pdf must be parsed before analysis can be performed on the content. It will often be formatted as xml but not always. In this case oddpub has a parser built in and it writes out the file when you run the tool.

**Comment by quang-ng** [comment]
Thank you @leej3 for your reply. 
> > Should we develop API endpoints like this, or simply create a script to run as we did before?
> 
> Yes the idea of encapsulating the pipeline as a service is to help isolate the dependencies across different pipelines. This one is not especially hard to run locally but we don't want to accumulate lots of conda environments.

From what I gather, we will set up a server to host the services, which will provide a POST request endpoint like `POST extract-metrics/`  with payload allow many PDF files, for processing PDF files. The API will invoke the function outlined in point 3. Is my understanding accurate?

**Comment by leej3** [comment]
> we will set up a server to host the services

For now we will keep things local. Python on whales offers a nice pythonic interface for managing docker (I used docker-compose [here](https://github.com/nimh-dsst/osm/blob/main/osm/_utils.py#L67-L100) but avoiding compose is fine too if you only have a single container to manage the lifecycle of).

> with payload allow many PDF files, for processing PDF files

Many files or one. You decide what works well.

> The API will invoke the function outlined in point 3. Is my understanding accurate?

Yes the script will use the service in 1. to create the outputs for the schema defined in 2.

It will probably be easiest to first create a conda environment with oddpub first to run it locally without using docker. That will allow you to  inspect the intermediates/output before working through the other steps.

**Comment by quang-ng** [comment]
Hello @leej3, for point 3.b, could you explain how Oddpub references  with Work, Document, Provenance,  RTransparentMetrics?

**Comment by leej3** [comment]
> for point 3.b, could you explain how Oddpub references with Work, Document, Provenance, RTransparentMetrics?

### Work/document

Previously "work" represented a publication of some sort but we didn't have an associated "document" (pdf). Now we have a "work" and also an associated "document". Multiple "documents" could back reference the same "work" but in practice until we come up with a strategy to deduplicate "works" it will be a 1-to-1 mapping between the two.

The table "RTransparentMetrics" above is a mistake... we have decided to stick to using Oddpub for now so "OddpubMetrics" will be the output table for this issue. Oddpub generates a subset of the fields that Rtransparent creates. "Provenance" captures information about the process of generating output from an input. So in this case the input is the pdf document and the output is the oddpub output.

**Comment by quang-ng** [comment]
PR for this issue: https://github.com/nimh-dsst/dsst-etl/pull/15
